#160577569 Enable users to reset their passwords

mutaimwiti opened the pull request:
**What does this PR do?**

Adds password reset logic. This is triggered by clicking forgot password link on the login or sign up pages. This allows the user to enter an email where the password reset link will be sent. After opening the link the user is taken to the frontend where they key in their new password.

**Description of Task to be completed?**

A user should be presented with a button and or link to reset their passwords on the SignUp/Register component.

**How should this be manually tested?**

This PR can be tested manually by following the following steps on the staging or review app:
1. Create an account.
2. Click Forgot password link on the signup or login form. Route: `/forgot-password`.
3. Enter email address and click send me link button.
4. Go to email and click the password reset link.
5. Enter the email, new password and confirm the password fields.
6. Click reset password button to reset the password. Route: `/reset-password`

**Any background context you want to provide?**

This PR requires the signup and login components to have the Forgot password links.

**What are the relevant pivotal tracker stories?**

[#160577569](https://www.pivotaltracker.com/story/show/160577569)

**Screenshots**
<img width="654" alt="screen shot 2018-11-14 at 20 22 39" src="https://user-images.githubusercontent.com/8180548/48501250-e2ccd280-e84d-11e8-8499-f6eeacf64eb6.png">
<img width="628" alt="screen shot 2018-11-14 at 20 23 24" src="https://user-images.githubusercontent.com/8180548/48501253-e52f2c80-e84d-11e8-9f3b-ee6e9de56019.png">
<img width="629" alt="screen shot 2018-11-14 at 20 27 24" src="https://user-images.githubusercontent.com/8180548/48501258-e6f8f000-e84d-11e8-9491-cd3558c22f1d.png">
<img width="748" alt="screen shot 2018-11-14 at 20 38 12" src="https://user-images.githubusercontent.com/8180548/48501260-e8c2b380-e84d-11e8-8b88-0e2748fc1e4d.png">
<img width="657" alt="screen shot 2018-11-14 at 20 44 17" src="https://user-images.githubusercontent.com/8180548/48501379-258eaa80-e84e-11e8-9fc4-38f9468658ff.png">

charisschomba reviewed (commented):
Fix hound flags.....otherwise it this starter skeleton is okay.

SrMoffat commented on src/pages/ForgotPasswordPage/index.js:
```diff
@@ -0,0 +1,18 @@
+class ResetPasswordPage extends Component {
+  render() {
+    return (
+      <>
```
Hi @mutaimwiti,
This is good work so far. A few things to address:
1. The [review app](https://ah-cd-frontend-staging-pr-14.herokuapp.com/) on `Heroku` is currently not working.
2. Small query from me, in this file `src/pages/ForgotPasswordPage/index.js`, for some reason I find line `4` very weird because we are importing `ForgotPasswordForm` from `ResetPasswordForm` or they are just similar components that you are reusing? Maybe we can think of a more generic name for the component in that case?

mutaimwiti replied:
@SrMoffat 
1. You need to access the `/forgot-password` route for you to access the UI.
2. My bad. I did it accidentally while creating the skeleton.

charisschomba reviewed (commented):
Nice work  @mutaimwiti on this PR but consider adding increasing the margin between navbar and forms.

paulrimiru commented on src/containers/ForgotPasswordForm/index.js:
```diff
@@ -0,0 +1,110 @@
+              <label className="active" htmlFor="email">Email</label>
+              {this.renderErrorMessage()}
+            </div>
+            <button type="submit" className="btn" disabled={isRequesting}>Send reset link</button>
```
Add these props into seperate lines

mutaimwiti replied:
Done

paulrimiru commented on src/containers/ForgotPasswordForm/index.js:
```diff
@@ -0,0 +1,110 @@
+        <div className="row">
+          <div className="col m10 offset-m1">
+            <p>{message}</p>
+            <button type="button" className="btn" onClick={this.handleResendLink}>Resend Link</button>
```
Put these props into separate lines

mutaimwiti replied:
Done

paulrimiru commented on src/containers/ForgotPasswordForm/state/reducer.js:
```diff
@@ -0,0 +1,26 @@
+      };
+    case FORGOT_PASSWORD_SUCCESS:
+      return {
+        ...state, ...action.payload, linkSent: true, isRequesting: false,
```
If the number of values you are accessing goes beyond three place them in separate lines

mutaimwiti replied:
Done

paulrimiru commented on src/containers/ResetPasswordForm/index.js:
```diff
@@ -0,0 +1,91 @@
+    confirmPassword: null,
+  };
+
+  handleUpdate = (ev) => {
```
Make the parameter `ev` more descriptive

mutaimwiti replied:
Done

paulrimiru commented on src/containers/ResetPasswordForm/index.js:
```diff
@@ -0,0 +1,91 @@
+        <div className="row">
+          <div className="col m10 offset-m1">
+            <div className="input-field">
+              <input type="email" name="email" className="validate" required onChange={this.handleUpdate} />
```
Add these to separate lines

mutaimwiti replied:
Done

paulrimiru commented on src/containers/ResetPasswordForm/index.js:
```diff
@@ -0,0 +1,91 @@
+              <label className="active" htmlFor="email">Email</label>
+            </div>
+            <div className="input-field">
+              <input type="password" name="password" className="validate" required onChange={this.handleUpdate} />
```
Add these to seperate lines

paulrimiru commented on src/containers/ResetPasswordForm/index.js:
```diff
@@ -0,0 +1,91 @@
+              <label className="active" htmlFor="password">New password</label>
+            </div>
+            <div className="input-field">
+              <input type="password" name="confirmPassword" className="validate" required onChange={this.handleUpdate} />
```
Same here

paulrimiru commented on src/containers/ResetPasswordForm/index.js:
```diff
@@ -0,0 +1,91 @@
+              <input type="password" name="confirmPassword" className="validate" required onChange={this.handleUpdate} />
+              <label className="active" htmlFor="confirmPassword">Confirm password</label>
+            </div>
+            <button type="submit" className="btn" disabled={isRequesting}>Reset password</button>
```
Same thing here

mutaimwiti replied:
Done

paulrimiru commented on src/containers/ResetPasswordForm/state/reducer.js:
```diff
@@ -0,0 +1,14 @@
+      return { ...state, isRequesting: true };
+    case RESET_PASSWORD_SUCCESS:
+      return {
+        ...state, ...action.payload, resetSuccessful: true, isRequesting: false,
```
Same thing here

mutaimwiti replied:
Done

paulrimiru commented on src/containers/ResetPasswordForm/state/actions.js:
```diff
@@ -0,0 +1,17 @@
+        payload: { message: res.data.message},
+      });
+    }).catch((error) => {
+    // dispatch fail action
```
Remove inline comments

mutaimwiti replied:
Done

paulrimiru commented on src/utils/api.js:
```diff
@@ -0,0 +1,53 @@
+import axios from 'axios';
+import config from './config';
+
+const token = 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpZCI6MSwidXNlcm5hbWUiOiJpbWlyZSIsImVtYWlsIjoiaW1pcmVAZ21haWwuY29tIiwiaWF0IjoxNTQxNTgyMjQzLCJleHAiOjE1NDE1ODk0NDN9.KqR2wO0Q94VjC3zr7CqwYYz-_6HrQY0j63jiBUILlcw';
```
Store this token in the environment variables

mutaimwiti replied:
Done

paulrimiru commented on src/containers/ForgotPasswordForm/index.js:
```diff
@@ -0,0 +1,110 @@
+import { connect } from 'react-redux';
+import { forgotPasswordAction, resendLinkAction } from './state/actions';
+
+
```
Remove extra line here

mutaimwiti replied:
Done

paulrimiru reviewed (changes requested):
Good work, make the requested changes

mutaimwiti commented:
@PaulKariukiRimiru I'm currently trying to fix hound issues. Its behaviour is inconsistent with the eslint config we are using. I also have a problem with my tests running locally but failing on Travis.

mutaimwiti commented:
Actually, I think Hound is passing at the moment. The failing tests are the impediment.

paulrimiru commented:
@mutaimwiti Fix the merge conflicts

mutaimwiti commented:
> @mutaimwiti Fix the merge conflicts

Working on it.

mutaimwiti committed:
feat(reset-password): enable users to reset passwords

- create forgot password send link form
- create forgot password resend link ui
- create forgot password page
- create reset password form
- create reset password success ui
- create forgot password page
- test forgot password form and page
- test reset password form and page

[Starts #160577569]

gitaumoses4 commented on src/routes/index.js:
```diff
@@ -76,6 +78,17 @@ export default () => (
         component={SignUpPage}
       />
       <Route
+      <Route
```
Check this. I believe it is the only remaining violation from hound CI.

mutaimwiti replied:
Done

gitaumoses4 reviewed (approved):
Looks Good To Me.

charisschomba commented on src/containers/ForgotPasswordForm/state/actions.test.js:
```diff
@@ -0,0 +1,63 @@
+import config from '../../../utils/config';
+import { forgotPasswordAction, resendLinkAction } from './actions';
+import {
+  FORGOT_PASSWORD_REQUEST, FORGOT_PASSWORD_FAIL, FORGOT_PASSWORD_SUCCESS, RESEND_RESET_LINK,
```
Import each of this imports in a new line for easy readability.

mutaimwiti replied:
Done

charisschomba commented on src/containers/ForgotPasswordForm/state/reducer.js:
```diff
@@ -0,0 +1,26 @@
+import {
+  FORGOT_PASSWORD_REQUEST, FORGOT_PASSWORD_FAIL, FORGOT_PASSWORD_SUCCESS, RESEND_RESET_LINK,
+} from './types';
```
The same  case applies here.

mutaimwiti replied:
Done

charisschomba commented on src/containers/ForgotPasswordForm/state/reducer.js:
```diff
@@ -0,0 +1,39 @@
+    default:
+      return state;
+  }
+};
```
Consider adding a new line after every case.

mutaimwiti replied:
Done

charisschomba commented on src/containers/ForgotPasswordForm/state/reducer.test.js:
```diff
@@ -0,0 +1,44 @@
+import reducer from './reducer';
+import {
+  FORGOT_PASSWORD_FAIL, FORGOT_PASSWORD_REQUEST, FORGOT_PASSWORD_SUCCESS, RESEND_RESET_LINK,
+} from './types';
```
Same here, each import in a new line

mutaimwiti replied:
Done

charisschomba commented on src/containers/ResetPasswordForm/index.js:
```diff
@@ -0,0 +1,139 @@
+import { bindActionCreators } from 'redux';
+import { connect } from 'react-redux';
+import { Link } from 'react-router-dom';
+import { resetPasswordAction, resetFieldErrorAction, resetGenericErrorAction } from './state/actions';
```
same case here.

mutaimwiti replied:
Done

charisschomba commented on src/containers/ResetPasswordForm/state/actions.test.js:
```diff
@@ -0,0 +1,78 @@
+import MockAdapter from 'axios-mock-adapter';
+import api from '../../../utils/api';
+import config from '../../../utils/config';
+import { resetFieldErrorAction, resetGenericErrorAction, resetPasswordAction } from './actions';
```
Add each import in a new line

mutaimwiti replied:
Done

charisschomba commented on src/pages/ForgotPasswordPage/index.js:
```diff
@@ -0,0 +1,40 @@
+
+export function ForgotPasswordPage({ isRequesting }) {
+  return (
+    <>
```
consider adding a div instead of this <>

mutaimwiti replied:
Noted

charisschomba commented on src/pages/ResetPasswordPage/index.js:
```diff
@@ -0,0 +1,39 @@
+import NavBar from "../../containers/NavBar";
+
+export const ResetPasswordPage = ({ isRequesting }) => (
+  <>
```
same case here

mutaimwiti replied:
Noted
